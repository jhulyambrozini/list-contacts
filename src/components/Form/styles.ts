import styled from 'styled-components'
import { colors, typography } from '../../styles'

export const FormContainer = styled.form`
  padding: 6rem 4rem 0 4rem;
  width: 100%;
`

export const FormInputsControls = styled.div`
  display: flex;
  gap: 40px;
  justify-content: center;
  margin-top: 11.6rem;

  > div {
    width: 300px;
  }

  @media screen and (max-width: 768px) {
    > div {
      width: 100%;
    }
  }

  @media screen and (max-width: 426px) {
    flex-direction: column;
    align-items: center;
    width: 100%;
    gap: 0;
  }
`
export const FormButtonsControls = styled.div`
  display: flex;
  justify-content: space-between;
`

export const PhotoContainer = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  gap: 8px;
`

export const Photo = styled.img`
  width: 100%;
  height: 100%;
  border-radius: 50%;
  object-fit: cover;
  position: absolute;
`

export const InputAddPhotoContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
  background-color: #fff;
  width: 130px;
  height: 130px;
  border-radius: 50%;

  input {
    opacity: 0;
    width: 100%;
    height: 100%;
    z-index: 9999;
  }

  img {
    position: absolute;
  }

  @media screen and (max-width: 426px) {
    height: 100px;
    width: 100px;
  }
`
export const FullName = styled.span`
  color: ${colors.primary09};
  text-align: center;

  font-family: ${typography.secondary};
  font-size: 24px;
  font-weight: 500;

  @media screen and (max-width: 426px) {
    span {
      font-size: 2rem;
    }
  }
`
