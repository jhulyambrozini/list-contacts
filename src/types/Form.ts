export interface ValuesForm {
  firstName: string
  lastName: string
  email: string
  tel: string
}
