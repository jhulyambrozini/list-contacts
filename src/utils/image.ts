const getRandomColor = (): string => {
  const red = Math.floor(Math.random() * 256)
  const green = Math.floor(Math.random() * 256)
  const blue = Math.floor(Math.random() * 256)

  return `rgb(${red},${green},${blue})`
}

export const generateRandomImage = (
  text: string,
  width: number,
  height: number
): string => {
  const canvas = document.createElement('canvas')
  canvas.width = width
  canvas.height = height

  const ctx = canvas.getContext('2d')
  if (!ctx) {
    throw new Error('Canvas 2D context is not supported')
  }

  // Preencher o canvas com uma cor aleatória
  ctx.fillStyle = getRandomColor()
  ctx.fillRect(0, 0, width, height)

  // Configurar o texto no centro
  ctx.fillStyle = 'white'
  ctx.font = '50px Arial'
  ctx.textAlign = 'center'
  ctx.textBaseline = 'middle'

  // Capitalizar a primeira letra da string e colocá-la no centro
  const firstLetter = text.charAt(0).toUpperCase()
  ctx.fillText(firstLetter, width / 2, height / 2)

  // Converter o canvas para um formato de dados URL
  const dataURL = canvas.toDataURL('image/png')
  return dataURL
}

export const validateImageFormat = (file: File): boolean => {
  const allowedFormats = ['image/png', 'image/jpeg']
  return allowedFormats.includes(file.type)
}
