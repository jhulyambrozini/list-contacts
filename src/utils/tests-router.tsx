import { ReactNode } from 'react'
import { BrowserRouter, useNavigate } from 'react-router-dom'
import { render } from '@testing-library/react'

const renderWithRouter = (component: ReactNode) => {
  let navigate: any
  const Wrapper = () => {
    navigate = useNavigate()
    return component
  }

  return {
    ...render(
      <BrowserRouter>
        <Wrapper />
      </BrowserRouter>
    ),
    navigate
  }
}

export default renderWithRouter
