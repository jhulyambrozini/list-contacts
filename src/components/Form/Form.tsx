import { ChangeEvent, FormEvent, useState } from "react"
import { Button, Field, FormStyle } from "./styles"
import { useDispatch } from "react-redux"
import { register } from "../../store/reducers/contacts"

const Form = () => {
  const dispatch = useDispatch()
  const [name, setName] = useState('')
  const [email, setEmail] = useState('')
  const [tel, setTel] = useState('')

  const createContact = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    if(name === '' || email === '' || tel === '') {
      alert('Preencha todos os campo')
    } else {
      dispatch(
        register({
          name,
          email,
          tel
        })
      )
    }
    console.log('submited', e)

  }

  return (
    <FormStyle onSubmit={createContact}>
        <Field type="text" placeholder="Nome" value={name} onChange={(e: ChangeEvent<HTMLInputElement>) => setName(e.target.value)}/>
        <Field type="number" placeholder="2799999-9999" value={tel} onChange={(e: ChangeEvent<HTMLInputElement>) => setTel(e.target.value)}/>
        <Field type="email" placeholder="exemple@email.com" value={email} onChange={(e: ChangeEvent<HTMLInputElement>) => setEmail(e.target.value)}/>
        <Button type="submit">SALVAR</Button>
    </FormStyle>
  )
}

export default Form
