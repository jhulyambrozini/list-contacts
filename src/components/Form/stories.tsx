import { Meta } from '@storybook/react'
import Form, { FormProps } from '.'
import mockStore from '../../mocks/redux-store'
import { Provider } from 'react-redux'

const meta = {
  title: 'components/Form',
  component: Form,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered'
  },
  args: {
    id: '1'
  }
} satisfies Meta<typeof Form>

export default meta

export const NewContact = () => {
  return <Form />
}

export const EditContact = (args: FormProps) => {
  const store = mockStore({
    contact: {
      items: [
        {
          id: '1',
          firstName: 'John',
          lastName: 'Doe',
          email: 'john@example.com',
          tel: '9999-99999',
          image: 'https://placehold.co/90x90/030020/png'
        }
      ]
    },
    popUp: {
      isPopupOpen: false
    }
  })
  return (
    <Provider store={store}>
      <Form {...args} />
    </Provider>
  )
}
