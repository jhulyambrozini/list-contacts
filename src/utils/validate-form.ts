import { FormikProps } from 'formik'
import { ValuesForm } from '../types/Form'

export const getFieldStatus = (
  field: string,
  form: FormikProps<ValuesForm>
) => {
  const isTouched = field in form.touched
  const isInvalid = field in form.errors
  return { isTouched, isInvalid }
}

export const getMessageError = (
  field: string,
  form: FormikProps<ValuesForm>,
  message?: string
) => {
  const { isTouched, isInvalid } = getFieldStatus(field, form)
  return isTouched && isInvalid ? message : false
}

export const getSuccess = (field: string, form: FormikProps<ValuesForm>) => {
  const { isTouched, isInvalid } = getFieldStatus(field, form)
  return isTouched && !isInvalid
}
