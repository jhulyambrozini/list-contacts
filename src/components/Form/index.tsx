import { ChangeEvent, useState } from 'react'
import { useFormik } from 'formik'
import { v4 as uuidv4 } from 'uuid'
import * as Yup from 'yup'
import { useDispatch, useSelector } from 'react-redux'
import { useNavigate } from 'react-router-dom'

import InputGroup from '../InputGroup'
import { Button } from '../Button'
import PopupConfirm from '../PopupComfirm'

import * as Style from './styles'
import trashIcom from '../../assets/trash-icon.svg'
import addPhotoIcon from '../../assets/addPhoto-icon.svg'

import { ValuesForm } from '../../types/Form'
import { TContact, TContactWithoutId } from '../../types/Contact'

import { generateRandomImage, validateImageFormat } from '../../utils/image'
import { getMessageError, getSuccess } from '../../utils/validate-form'

import { edit, register } from '../../store/reducers/contacts'
import { RootState } from '../../store'
import { openPopup } from '../../store/reducers/popup'

export type FormProps = {
  id?: string
}

const Form = ({ id }: FormProps) => {
  const dispatch = useDispatch()
  const navigate = useNavigate()

  const { items } = useSelector((state: RootState) => state.contact)
  const contactInfos = items.find((i: TContact) => i.id === id)

  const [image, setImage] = useState<string | undefined>(
    contactInfos?.image || ''
  )

  const form = useFormik<ValuesForm>({
    initialValues: {
      firstName: contactInfos?.firstName || '',
      lastName: contactInfos?.lastName || '',
      email: contactInfos?.email || '',
      tel: contactInfos?.tel || ''
    },
    validationSchema: Yup.object({
      firstName: Yup.string().required('O campo é obrigatório'),
      lastName: Yup.string().required('O campo é obrigatório'),
      tel: Yup.string().required('O campo é obrigatório'),
      email: Yup.string()
        .email('Digite um email válido')
        .required('O campo é obrigatório')
    }),
    onSubmit: async (values) => {
      if (image) {
        const data = {
          ...values,
          image
        }
        await registerData(data)
      } else {
        const randomImage = generateRandomImage(form.values.firstName, 100, 100)
        const data = {
          ...values,
          image: randomImage
        }
        await registerData(data)
      }
    }
  })

  const handleImageChange = (event: ChangeEvent<HTMLInputElement>) => {
    const input = event.target
    const file = input.files?.[0]

    if (file && validateImageFormat(file)) {
      const reader = new FileReader()

      reader.onload = (e) => {
        setImage(e.target?.result as string)
      }

      reader.readAsDataURL(file)
    } else {
      setImage('')
    }
  }

  const updateContact = async (values: TContactWithoutId) => {
    if (id) {
      const data = {
        ...values,
        id
      }
      await dispatch(edit(data))
      navigate('/')
    }
  }

  const registerNewContact = async (values: TContactWithoutId) => {
    if (isContactExist()) {
      alert(
        `Este telefone: ${form.values.tel} ou email: ${form.values.email} já está na agenda`
      )
    } else {
      const data = {
        ...values,
        id: uuidv4()
      }
      await dispatch(register(data))
      navigate('/')
    }
  }

  const registerData = async (values: TContactWithoutId) => {
    if (id) {
      await updateContact(values)
    } else {
      await registerNewContact(values)
    }
  }

  const isContactExist = () => {
    for (let contact of items) {
      const emailExists = contact.email === form.values.email
      const telExists = contact.tel === form.values.tel

      if (emailExists || telExists) return true
    }

    return false
  }

  const getFullName = () => {
    if (contactInfos?.firstName && contactInfos.lastName) {
      return `${contactInfos.firstName} ${contactInfos.lastName}`
    } else if (form.values.firstName !== '' && form.values.lastName !== '') {
      return `${form.values.firstName} ${form.values.lastName}`
    }
    return 'Insira um nome'
  }

  return (
    <Style.FormContainer onSubmit={form.handleSubmit} data-testid="form">
      <Style.PhotoContainer>
        <Style.InputAddPhotoContainer aria-label="Input image">
          <input
            type="file"
            onChange={handleImageChange}
            accept="image/png, image/jpeg"
            multiple={false}
            title="Adicionar foto"
          />
          {image ? (
            <Style.Photo src={image} alt="Imagem de perfil" />
          ) : (
            <img src={addPhotoIcon} alt="Icone de uma câmera" />
          )}
        </Style.InputAddPhotoContainer>
        <Style.FullName>{getFullName()}</Style.FullName>
      </Style.PhotoContainer>

      <Style.FormInputsControls>
        <div>
          <InputGroup
            id="firstName"
            label="Nome"
            type="text"
            onBlur={form.handleBlur}
            onChange={form.handleChange}
            value={form.values.firstName}
            erroMessage={getMessageError(
              'firstName',
              form,
              form.errors.firstName
            )}
            isSuccess={getSuccess('firstName', form)}
          />
          <InputGroup
            id="lastName"
            label="Sobrenome"
            type="text"
            onBlur={form.handleBlur}
            onChange={form.handleChange}
            value={form.values.lastName}
            erroMessage={getMessageError(
              'lastName',
              form,
              form.errors.lastName
            )}
            isSuccess={getSuccess('lastName', form)}
          />
        </div>

        <div>
          <InputGroup
            id="tel"
            label="Telefone"
            type="tel"
            onBlur={form.handleBlur}
            onChange={form.handleChange}
            value={form.values.tel}
            erroMessage={getMessageError('tel', form, form.errors.tel)}
            isSuccess={getSuccess('tel', form)}
          />
          <InputGroup
            id="email"
            label="Email"
            type="email"
            onBlur={form.handleBlur}
            onChange={form.handleChange}
            value={form.values.email}
            erroMessage={getMessageError('email', form, form.errors.email)}
            isSuccess={getSuccess('email', form)}
          />
        </div>
      </Style.FormInputsControls>

      <Style.FormButtonsControls>
        {id && (
          <Button.Circle
            type="button"
            padding="1.6rem"
            onclick={() => dispatch(openPopup())}
            title="Excluir contato"
          >
            <Button.Icon alt="Icone de lixeira (excluir)" icon={trashIcom} />
          </Button.Circle>
        )}

        <Button.Secondary
          padding="1.2rem 2.4rem"
          type="submit"
          title="Salvar contato"
        >
          <Button.Label label="SALVAR" />
        </Button.Secondary>
      </Style.FormButtonsControls>

      {id && <PopupConfirm id={id} />}
    </Style.FormContainer>
  )
}

export default Form
