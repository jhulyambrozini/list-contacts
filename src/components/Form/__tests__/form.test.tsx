import { fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { renderWithProvider } from '../../../utils/tests-redux'
import { MemoryRouter } from 'react-router-dom'
import Form from '..'
import userEvent from '@testing-library/user-event'
import { vi } from 'vitest'

describe('<Form />', () => {
  // deve exluir contato quando aperta no botão de excluir

  it('deve renderizar apenas o botão de salvar na pagina de newContact', () => {
    const { container } = renderWithProvider(
      <MemoryRouter>
        <Form />
      </MemoryRouter>
    )

    expect(screen.getAllByRole('button').length).toEqual(1)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('deve preencher o formulario corretamente e registrar', async () => {
    const handleSubmit = vi.fn()

    renderWithProvider(
      <MemoryRouter>
        <Form />
      </MemoryRouter>
    )

    const inputFirstName = screen.getByRole('textbox', {
      name: 'Nome'
    })
    const inputLastName = screen.getByRole('textbox', {
      name: /sobrenome/i
    })
    const inputPhone = screen.getByRole('textbox', {
      name: /telefone/i
    })
    const inputEmail = screen.getByRole('textbox', {
      name: /email/i
    })

    const inputImage = screen.getByTitle('Adicionar foto')

    const file = new File(['(image)'], 'foto.png', {
      type: 'image/png'
    })

    fireEvent.change(inputFirstName, { target: { value: 'kamilly' } })
    fireEvent.change(inputLastName, { target: { value: 'silva' } })
    fireEvent.change(inputEmail, { target: { value: 'exemple@gmail.com' } })
    fireEvent.change(inputPhone, { target: { value: '0000-0000' } })
    userEvent.upload(inputImage, file)

    const register = screen.getByTestId('form')
    register.onsubmit = handleSubmit

    fireEvent.submit(register)

    expect(handleSubmit).toHaveBeenCalled()
  })

  it('deve renderizar o formulario de edição quando receber um id props ', () => {
    renderWithProvider(
      <MemoryRouter>
        <Form id="1" />
      </MemoryRouter>,
      {
        preloadedState: {
          contact: {
            items: [
              {
                email: 'exemple@gmail.com',
                id: '1',
                image: '/img.png',
                firstName: 'Kamilly',
                lastName: 'Silva',
                tel: '11 99999-9999'
              }
            ]
          }
        }
      }
    )

    const fullName = screen.getByText('Kamilly Silva')
    expect(fullName).toBeInTheDocument()
  })
  it('deve salvar a edição quando clicar no botão salvar', () => {
    const handleSubmitEdit = vi.fn()
    renderWithProvider(
      <MemoryRouter>
        <Form id="1" />
      </MemoryRouter>,
      {
        preloadedState: {
          contact: {
            items: [
              {
                email: 'exemple@gmail.com',
                id: '1',
                image: '/img.png',
                firstName: 'Kamilly',
                lastName: 'Silva',
                tel: '11 99999-9999'
              }
            ]
          }
        }
      }
    )
    const inputFirstName = screen.getByRole('textbox', {
      name: 'Nome'
    })
    fireEvent.change(inputFirstName, { target: { value: 'kamilla' } })

    const saveButton = screen.getByRole('button', { name: /salvar/i })

    saveButton.onclick = handleSubmitEdit

    fireEvent.click(saveButton)

    expect(handleSubmitEdit).toHaveBeenCalled()
  })
})
