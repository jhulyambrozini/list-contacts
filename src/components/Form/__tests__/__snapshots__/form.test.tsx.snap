// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Form /> > should only render the save button on the newContact page 1`] = `
<form
  class="sc-gKHVLF cgYcjW"
  data-testid="form"
>
  <div
    class="sc-pqitP inGyiQ"
  >
    <div
      aria-label="Input image"
      class="sc-kDnyCx dvnzHB"
    >
      <input
        accept="image/png, image/jpeg"
        title="Adicionar foto"
        type="file"
      />
      <img
        alt="Icone de uma cÃ¢mera"
        src="/src/assets/addPhoto-icon.svg"
      />
    </div>
    <span
      class="sc-ipMvLY wVEqB"
    >
      Insira um nome
    </span>
  </div>
  <div
    class="sc-iBAaJG guBMRz"
  >
    <div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="firstName"
        >
          Nome
        </label>
        <input
          class=""
          id="firstName"
          type="text"
          value=""
        />
        <span />
      </div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="lastName"
        >
          Sobrenome
        </label>
        <input
          class=""
          id="lastName"
          type="text"
          value=""
        />
        <span />
      </div>
    </div>
    <div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="tel"
        >
          Telefone
        </label>
        <input
          class=""
          id="tel"
          type="tel"
          value=""
        />
        <span />
      </div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          class=""
          id="email"
          type="email"
          value=""
        />
        <span />
      </div>
    </div>
  </div>
  <div
    class="sc-ftLKQv edMseL"
  >
    <button
      class="sc-grXZZQ sc-hLclGa lcpiBB hVEQyo"
      padding="1.2rem 2.4rem"
      title="Salvar contato"
      type="submit"
    >
      <span>
        SALVAR
      </span>
    </button>
  </div>
</form>
`;

exports[`<Form /> > should render the edit form when receiving an id props 1`] = `
<form
  class="sc-gKHVLF cgYcjW"
  data-testid="form"
>
  <div
    class="sc-pqitP inGyiQ"
  >
    <div
      aria-label="Input image"
      class="sc-kDnyCx dvnzHB"
    >
      <input
        accept="image/png, image/jpeg"
        title="Adicionar foto"
        type="file"
      />
      <img
        alt="Imagem de perfil"
        class="sc-jrkPcm rBxP"
        src="/img.png"
      />
    </div>
    <span
      class="sc-ipMvLY wVEqB"
    >
      Kamilly Silva
    </span>
  </div>
  <div
    class="sc-iBAaJG guBMRz"
  >
    <div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="firstName"
        >
          Nome
        </label>
        <input
          class=""
          id="firstName"
          type="text"
          value="Kamilly"
        />
        <span />
      </div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="lastName"
        >
          Sobrenome
        </label>
        <input
          class=""
          id="lastName"
          type="text"
          value="Silva"
        />
        <span />
      </div>
    </div>
    <div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="tel"
        >
          Telefone
        </label>
        <input
          class=""
          id="tel"
          type="tel"
          value="11 99999-9999"
        />
        <span />
      </div>
      <div
        aria-label="Input group"
        class="sc-bcPKhP eBMfoZ"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          class=""
          id="email"
          type="email"
          value="exemple@gmail.com"
        />
        <span />
      </div>
    </div>
  </div>
  <div
    class="sc-ftLKQv edMseL"
  >
    <button
      class="sc-grXZZQ sc-eDnVMP gCOyb dzJcPW"
      padding="1.6rem"
      title="Excluir contato"
      type="button"
    >
      <img
        alt="Icone de lixeira (excluir)"
        src="/src/assets/trash-icon.svg"
      />
    </button>
    <button
      class="sc-grXZZQ sc-hLclGa lcpiBB hVEQyo"
      padding="1.2rem 2.4rem"
      title="Salvar contato"
      type="submit"
    >
      <span>
        SALVAR
      </span>
    </button>
  </div>
</form>
`;
